src/python/pants/backend/python/rules/run_setup_py.py

```diff
@@ -49,12 +49,121 @@
 from pants.source.source_root import SourceRootConfig
 
 
+class NoOwnerError(Exception):
+  """Indicates an exportable target has no owning exported target."""
+
+
+class AmbiguousOwnerError(Exception):
+  """Indicates an exportable target has more than one owning exported target."""
+
+
+@dataclass(frozen=True)
+class ExportedTarget:
+  """A target that explicitly exports an artifact, using a `provides=` stanza.
+
+  The code provided by this artifact can be from this target or from any targets it owns.
+  """
+  hydrated_target: HydratedTarget
+
+
+@dataclass(frozen=True)
+class DependencyOwner:
+  """An ExportedTarget in its role as an owner of other targets.
+
+  We need this type to prevent rule ambiguities when computing the list of targets
+  owned by an ExportedTarget (which involves going from ExportedTarget -> dep -> owner (which
+  is itself an ExportedTarget) and checking if owner is this the original ExportedTarget.
+  """
+  exported_target: ExportedTarget
+
+
+@dataclass(frozen=True)
+class OwnedDependency:
+  """A target that is owned by some ExportedTarget.
+
+  Code in this target is published in the owner's artifact.
+
+  The owner of a target T is T's closest filesystem ancestor among the exported targets
+  that directly or indirectly depend on it (including T itself).
+  """
+  hydrated_target: HydratedTarget
+
+
+class OwnedDependencies(Collection[OwnedDependency]):
+  pass
+
+
 @dataclass(frozen=True)
 class SetuptoolsSetup:
   """The setuptools tool."""
   requirements_pex: Pex
 
 
+def _is_exported(target: HydratedTarget) -> bool:
+  return getattr(target.adaptor, 'provides', None) is not None
+
+
+@rule(name="Get owned targets")
+async def get_owned_dependencies(dependency_owner: DependencyOwner) -> OwnedDependencies:
+  """Find the dependencies of dependency_owner that are owned by it.
+
+  Includes dependency_owner itself.
+  """
+  tht = await Get[TransitiveHydratedTargets](
+    BuildFileAddresses([dependency_owner.exported_target.hydrated_target.address]))
+  all_tgts = list(tht.closure)
```

Comment:
I think that was left over from a previous iteration, when I needed a stable order. Now I actually don't. 

But generally, I don't know how one can be sure that `TransitiveHydatedTargets.closure` is an OrderedSet. The type annotation is `Any`.  So nothing formal appears to guarantee stable order.

Reply:
It’s only Any because we haven’t yet added proper annotations. Also, we can’t use typed OrderedSets yet. We can declare it’s an OrderedSet but not what the element type is. 

Fwit, in my free time I put up a PR to the ordered-set library to improve distribution of their type stubs. If that lands, I’ll put up a PR to replace Twitter Commons version with this library.